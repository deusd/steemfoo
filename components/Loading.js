import React from 'react'
import { ActivityIndicator, View, Text } from 'react-native'
import Row from './layouts/Row'

const Loading = () => (
  <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center' }}>
    <Row>
      <Text style={{ marginRight: 10 }}>Loading posts</Text>
      <ActivityIndicator />
    </Row>
  </View>
)

export default Loading
