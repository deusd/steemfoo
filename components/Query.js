// @flow
import React from 'react'
import { Query as GQLQuery } from 'react-apollo'
import { Text, Button, View } from 'react-native'
import Loading from '../components/Loading'

class QueryRender extends React.Component<{
  loading: boolean,
  error: any,
  data: any,
  refetch: Promise,
  children: Function,
}> {
  state = {
    refetching: false,
  }

  unsetRefetch() {
    this.setState({ refetching: false })
  }

  handleRefetch = refetch => () => {
    this.setState({ refetching: true })

    refetch()
      .then(this.unsetRefetch)
      .catch(this.unsetRefetch)
  }

  render() {
    const { loading, error, data, refetch } = this.props

    if (loading || this.state.refetching) return <Loading />
    if (error)
      return (
        <View
          style={{
            flex: 1,
            justifyContent: 'center',
            alignItems: 'center',
          }}
          testID={'ErrorWrapper'}
        >
          <Text>Error :(</Text>
          <Button title={'Retry'} onPress={this.handleRefetch(refetch)} />
        </View>
      )

    return <View>{this.props.children(data)}</View>
  }
}

export default QueryRender
