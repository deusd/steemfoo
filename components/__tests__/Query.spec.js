import 'react-native'
import { shallow } from 'enzyme'
import React from 'react'
import { View } from 'react-native'
import gql from 'graphql-tag'
import Query from '../Query'
import Loading from '../Loading'

const query = gql`
  {
    posts(postType: TRENDING, limit: 1) {
      id
      title
    }
  }
`

describe('Query component', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(
      <Query query={query}>
        {data => <View testID="TestComponent">{data}</View>}
      </Query>
    )
  })

  it('should render', () => {
    expect(wrapper.shallow()).toMatchSnapshot()
  })

  describe('when loading', () => {
    it('should show the loading component', () => {
      wrapper.setProps({ loading: true })
      expect(wrapper.contains(<Loading />)).toBe(true)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when refetching', () => {
    it('should show the loading component', () => {
      wrapper.setState({ refetching: true })
      expect(wrapper.contains(<Loading />)).toBe(true)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when not loading', () => {
    it('should not show the loading component', () => {
      wrapper.setState({ refetching: false })
      wrapper.setProps({ loading: false })
      expect(wrapper.contains(<Loading />)).toBe(false)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when an error occurs', () => {
    it('should show the error', () => {
      wrapper.setProps({ error: true })
      wrapper.setState({ refetching: false })
      expect(wrapper.props().testID).toBe('ErrorWrapper')
    })

    it('should should stop refetching when promise resolves', done => {
      const refetch = jest.fn(() => Promise.resolve())
      wrapper.setProps({ error: true, refetch })
      wrapper
        .find({ testID: 'RefetchButton' })
        .props()
        .onPress()
      setImmediate(() => {
        expect(wrapper.state('refetching')).toBe(false)
        done()
      })
    })

    it('should should stop refetching when promise rejects', done => {
      const refetch = jest.fn(() => Promise.reject())
      wrapper.setProps({ error: true, refetch })
      wrapper
        .find({ testID: 'RefetchButton' })
        .props()
        .onPress()
      setImmediate(() => {
        expect(wrapper.state('refetching')).toBe(false)
        done()
      })
    })
  })

  describe('when loaded with data', () => {
    it('should render with child when data loads', () => {
      const data = 'testData'
      wrapper.setProps({ data })
      expect(
        wrapper
          .find({ testID: 'TestComponent' })
          .render()
          .text()
      ).toBe('testData')
    })
  })
})
