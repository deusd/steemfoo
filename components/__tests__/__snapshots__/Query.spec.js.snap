// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query component should render 1`] = `
<View>
  <Component
    testID="TestComponent"
  />
  <Button
    onPress={[Function]}
    testID="CrashButton"
    title="Crash"
  />
</View>
`;

exports[`Query component when loading should show the loading component 1`] = `<Loading />`;

exports[`Query component when not loading should not show the loading component 1`] = `
<Component>
  <Component
    testID="TestComponent"
  />
  <Button
    onPress={[Function]}
    testID="CrashButton"
    title="Crash"
  />
</Component>
`;

exports[`Query component when refetching should show the loading component 1`] = `<Loading />`;
